File: permissions/bitfield_test.ts
import { strictEqual } from "node:assert";
import { describe, test } from "node:test";
import * as permissions from "./bitfield.ts";

describe("has", () => {
  const tests: { name: string; base: bigint; check: bigint; want: boolean }[] = [
    { name: "single bit set", base: permissions.ViewChannels, check: permissions.ViewChannels, want: true },
    { name: "single bit unset", base: 0n, check: permissions.ViewChannels, want: false },
    {
      name: "multiple bits all set",
      base: permissions.ViewChannels | permissions.SendMessages,
      check: permissions.ViewChannels | permissions.SendMessages,
      want: true,
    },
    {
      name: "multiple bits partial",
      base: permissions.ViewChannels,
      check: permissions.ViewChannels | permissions.SendMessages,
      want: false,
    },
    { name: "zero check always true", base: 0n, check: 0n, want: true },
    {
      name: "all permissions has every bit",
      base: permissions.AllPermissions,
      check: permissions.ManageServer,
      want: true,
    },
    {
      name: "all permissions has voice",
      base: permissions.AllPermissions,
      check: permissions.VoiceConnect | permissions.VoiceSpeak,
      want: true,
    },
    {
      name: "all permissions has moderation",
      base: permissions.AllPermissions,
      check: permissions.ManageAutomod | permissions.SlowmodeExempt,
      want: true,
    },
  ];

  for (const tt of tests) {
    test(tt.name, () => {
      strictEqual(permissions.has(tt.base, tt.check), tt.want);
    });
  }
});

describe("add", () => {
  const tests: { name: string; base: bigint; add: bigint; want: bigint }[] = [
    { name: "add to zero", base: 0n, add: permissions.ViewChannels, want: permissions.ViewChannels },
    {
      name: "add existing is idempotent",
      base: permissions.ViewChannels,
      add: permissions.ViewChannels,
      want: permissions.ViewChannels,
    },
    {
      name: "combine two",
      base: permissions.ViewChannels,
      add: permissions.SendMessages,
      want: permissions.ViewChannels | permissions.SendMessages,
    },
    {
      name: "add to complex",
      base: permissions.ViewChannels | permissions.ManageServer,
      add: permissions.BanMembers,
      want: permissions.ViewChannels | permissions.ManageServer | permissions.BanMembers,
    },
  ];

  for (const tt of tests) {
    test(tt.name, () => {
      strictEqual(permissions.add(tt.base, tt.add), tt.want);
    });
  }
});

describe("remove", () => {
  const tests: { name: string; base: bigint; remove: bigint; want: bigint }[] = [
    {
      name: "remove present bit",
      base: permissions.ViewChannels | permissions.SendMessages,
      remove: permissions.SendMessages,
      want: permissions.ViewChannels,
    },
    {
      name: "remove absent bit is no-op",
      base: permissions.ViewChannels,
      remove: permissions.SendMessages,
      want: permissions.ViewChannels,
    },
    { name: "remove from zero", base: 0n, remove: permissions.ViewChannels, want: 0n },
    {
      name: "remove all",
      base: permissions.ViewChannels | permissions.SendMessages,
      remove: permissions.ViewChannels | permissions.SendMessages,
      want: 0n,
    },
  ];

  for (const tt of tests) {
    test(tt.name, () => {
      strictEqual(permissions.remove(tt.base, tt.remove), tt.want);
    });
  }
});

describe("AllPermissions", () => {
  const all = [
    // General
    permissions.ViewChannels,
    permissions.ManageChannels,
    permissions.ManageCategories,
    permissions.ManageRoles,
    permissions.ManageServer,
    permissions.ManageInvites,
    permissions.CreateInvites,
    permissions.ViewAuditLog,
    permissions.ManageWebhooks,
    permissions.ManageEmoji,
    // Membership
    permissions.KickMembers,
    permissions.BanMembers,
    permissions.TimeoutMembers,
    permissions.ChangeNicknames,
    permissions.ManageNicknames,
    permissions.AssignRoles,
    // Text
    permissions.SendMessages,
    permissions.SendMessagesThreads,
    permissions.CreateThreads,
    permissions.EmbedLinks,
    permissions.AttachFiles,
    permissions.AddReactions,
    permissions.MentionEveryone,
    permissions.ManageMessages,
    permissions.ReadMessageHistory,
    permissions.UseCommands,
    permissions.UseExternalEmoji,
    // Voice
    permissions.VoiceConnect,
    permissions.VoiceSpeak,
    permissions.VoiceVideo,
    permissions.VoiceScreenshare,
    permissions.VoicePTT,
    permissions.VoiceMuteMembers,
    permissions.VoiceDeafenMembers,
    permissions.VoiceMoveMembers,
    permissions.VoicePriority,
    // Moderation
    permissions.ManageAutomod,
    permissions.ViewReports,
    permissions.ManageReports,
    permissions.SlowmodeExempt,
  ];

  test("covers every bit", () => {
    for (const perm of all) {
      strictEqual(permissions.has(permissions.AllPermissions, perm), true, `AllPermissions missing bit ${perm}`);
    }
  });

  test("union of all constants equals AllPermissions", () => {
    let union = 0n;
    for (const perm of all) {
      union = permissions.add(union, perm);
    }
    strictEqual(union, permissions.AllPermissions);
  });
});

test("bits are contiguous", () => {
  for (let i = 0; i < 40; i++) {
    const bit = 1n << BigInt(i);
    strictEqual(permissions.has(permissions.AllPermissions, bit), true, `AllPermissions missing bit ${i}`);
  }
});

test("no overlapping bits", () => {
  const perms: { name: string; perm: bigint }[] = [
    { name: "ViewChannels", perm: permissions.ViewChannels },
    { name: "ManageChannels", perm: permissions.ManageChannels },
    { name: "ManageCategories", perm: permissions.ManageCategories },
    { name: "ManageRoles", perm: permissions.ManageRoles },
    { name: "ManageServer", perm: permissions.ManageServer },
    { name: "ManageInvites", perm: permissions.ManageInvites },
    { name: "CreateInvites", perm: permissions.CreateInvites },
    { name: "ViewAuditLog", perm: permissions.ViewAuditLog },
    { name: "ManageWebhooks", perm: permissions.ManageWebhooks },
    { name: "ManageEmoji", perm: permissions.ManageEmoji },
    { name: "KickMembers", perm: permissions.KickMembers },
    { name: "BanMembers", perm: permissions.BanMembers },
    { name: "TimeoutMembers", perm: permissions.TimeoutMembers },
    { name: "ChangeNicknames", perm: permissions.ChangeNicknames },
    { name: "ManageNicknames", perm: permissions.ManageNicknames },
    { name: "AssignRoles", perm: permissions.AssignRoles },
    { name: "SendMessages", perm: permissions.SendMessages },
    { name: "SendMessagesThreads", perm: permissions.SendMessagesThreads },
    { name: "CreateThreads", perm: permissions.CreateThreads },
    { name: "EmbedLinks", perm: permissions.EmbedLinks },
    { name: "AttachFiles", perm: permissions.AttachFiles },
    { name: "AddReactions", perm: permissions.AddReactions },
    { name: "MentionEveryone", perm: permissions.MentionEveryone },
    { name: "ManageMessages", perm: permissions.ManageMessages },
    { name: "ReadMessageHistory", perm: permissions.ReadMessageHistory },
    { name: "UseCommands", perm: permissions.UseCommands },
    { name: "UseExternalEmoji", perm: permissions.UseExternalEmoji },
    { name: "VoiceConnect", perm: permissions.VoiceConnect },
    { name: "VoiceSpeak", perm: permissions.VoiceSpeak },
    { name: "VoiceVideo", perm: permissions.VoiceVideo },
    { name: "VoiceScreenshare", perm: permissions.VoiceScreenshare },
    { name: "VoicePTT", perm: permissions.VoicePTT },
    { name: "VoiceMuteMembers", perm: permissions.VoiceMuteMembers },
    { name: "VoiceDeafenMembers", perm: permissions.VoiceDeafenMembers },
    { name: "VoiceMoveMembers", perm: permissions.VoiceMoveMembers },
    { name: "VoicePriority", perm: permissions.VoicePriority },
    { name: "ManageAutomod", perm: permissions.ManageAutomod },
    { name: "ViewReports", perm: permissions.ViewReports },
    { name: "ManageReports", perm: permissions.ManageReports },
    { name: "SlowmodeExempt", perm: permissions.SlowmodeExempt },
  ];

  for (let i = 0; i < perms.length; i++) {
    for (let j = i + 1; j < perms.length; j++) {
      strictEqual((perms[i].perm & perms[j].perm) === 0n, true, `${perms[i].name} and ${perms[j].name} share bits`);
    }
  }
});

test("deny wins pattern", () => {
  const base = permissions.ViewChannels | permissions.SendMessages | permissions.AddReactions;
  const roleAllow = permissions.ManageMessages;
  const roleDeny = permissions.SendMessages;

  const result = permissions.remove(permissions.add(base, roleAllow), roleDeny);

  strictEqual(permissions.has(result, permissions.SendMessages), false, "SendMessages should be denied");
  strictEqual(permissions.has(result, permissions.ViewChannels), true, "ViewChannels should remain");
  strictEqual(permissions.has(result, permissions.ManageMessages), true, "ManageMessages should be added");
  strictEqual(permissions.has(result, permissions.AddReactions), true, "AddReactions should remain untouched");
});

File: permissions/bitfield_test.go
package permissions

import "testing"

func TestHas(t *testing.T) {
	tests := []struct {
		name  string
		base  Permission
		check Permission
		want  bool
	}{
		{"single bit set", ViewChannels, ViewChannels, true},
		{"single bit unset", 0, ViewChannels, false},
		{"multiple bits all set", ViewChannels | SendMessages, ViewChannels | SendMessages, true},
		{"multiple bits partial", ViewChannels, ViewChannels | SendMessages, false},
		{"zero check always true", 0, 0, true},
		{"all permissions has every bit", AllPermissions, ManageServer, true},
		{"all permissions has voice", AllPermissions, VoiceConnect | VoiceSpeak, true},
		{"all permissions has moderation", AllPermissions, ManageAutomod | SlowmodeExempt, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := tt.base.Has(tt.check); got != tt.want {
				t.Errorf("Permission(%d).Has(%d) = %v, want %v", tt.base, tt.check, got, tt.want)
			}
		})
	}
}

func TestAdd(t *testing.T) {
	tests := []struct {
		name string
		base Permission
		add  Permission
		want Permission
	}{
		{"add to zero", 0, ViewChannels, ViewChannels},
		{"add existing is idempotent", ViewChannels, ViewChannels, ViewChannels},
		{"combine two", ViewChannels, SendMessages, ViewChannels | SendMessages},
		{"add to complex", ViewChannels | ManageServer, BanMembers, ViewChannels | ManageServer | BanMembers},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := tt.base.Add(tt.add); got != tt.want {
				t.Errorf("Permission(%d).Add(%d) = %d, want %d", tt.base, tt.add, got, tt.want)
			}
		})
	}
}

func TestRemove(t *testing.T) {
	tests := []struct {
		name   string
		base   Permission
		remove Permission
		want   Permission
	}{
		{"remove present bit", ViewChannels | SendMessages, SendMessages, ViewChannels},
		{"remove absent bit is no-op", ViewChannels, SendMessages, ViewChannels},
		{"remove from zero", 0, ViewChannels, 0},
		{"remove all", ViewChannels | SendMessages, ViewChannels | SendMessages, 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := tt.base.Remove(tt.remove); got != tt.want {
				t.Errorf("Permission(%d).Remove(%d) = %d, want %d", tt.base, tt.remove, got, tt.want)
			}
		})
	}
}

func TestAllPermissionsCoversEveryBit(t *testing.T) {
	all := []Permission{
		// General
		ViewChannels, ManageChannels, ManageCategories, ManageRoles, ManageServer,
		ManageInvites, CreateInvites, ViewAuditLog, ManageWebhooks, ManageEmoji,
		// Membership
		KickMembers, BanMembers, TimeoutMembers, ChangeNicknames, ManageNicknames, AssignRoles,
		// Text
		SendMessages, SendMessagesThreads, CreateThreads, EmbedLinks, AttachFiles,
		AddReactions, MentionEveryone, ManageMessages, ReadMessageHistory, UseCommands, UseExternalEmoji,
		// Voice
		VoiceConnect, VoiceSpeak, VoiceVideo, VoiceScreenshare, VoicePTT,
		VoiceMuteMembers, VoiceDeafenMembers, VoiceMoveMembers, VoicePriority,
		// Moderation
		ManageAutomod, ViewReports, ManageReports, SlowmodeExempt,
	}

	for _, perm := range all {
		if !AllPermissions.Has(perm) {
			t.Errorf("AllPermissions missing bit %d", perm)
		}
	}

	// Verify the union of all individual constants equals AllPermissions.
	var union Permission
	for _, perm := range all {
		union = union.Add(perm)
	}
	if union != AllPermissions {
		t.Errorf("union of all constants = %d, want AllPermissions = %d", union, AllPermissions)
	}
}

func TestBitsAreContiguous(t *testing.T) {
	for i := 0; i < 40; i++ {
		bit := Permission(1 << i)
		if !AllPermissions.Has(bit) {
			t.Errorf("AllPermissions missing bit %d", i)
		}
	}
}

func TestNoOverlappingBits(t *testing.T) {
	perms := []struct {
		name string
		perm Permission
	}{
		{"ViewChannels", ViewChannels},
		{"ManageChannels", ManageChannels},
		{"ManageCategories", ManageCategories},
		{"ManageRoles", ManageRoles},
		{"ManageServer", ManageServer},
		{"ManageInvites", ManageInvites},
		{"CreateInvites", CreateInvites},
		{"ViewAuditLog", ViewAuditLog},
		{"ManageWebhooks", ManageWebhooks},
		{"ManageEmoji", ManageEmoji},
		{"KickMembers", KickMembers},
		{"BanMembers", BanMembers},
		{"TimeoutMembers", TimeoutMembers},
		{"ChangeNicknames", ChangeNicknames},
		{"ManageNicknames", ManageNicknames},
		{"AssignRoles", AssignRoles},
		{"SendMessages", SendMessages},
		{"SendMessagesThreads", SendMessagesThreads},
		{"CreateThreads", CreateThreads},
		{"EmbedLinks", EmbedLinks},
		{"AttachFiles", AttachFiles},
		{"AddReactions", AddReactions},
		{"MentionEveryone", MentionEveryone},
		{"ManageMessages", ManageMessages},
		{"ReadMessageHistory", ReadMessageHistory},
		{"UseCommands", UseCommands},
		{"UseExternalEmoji", UseExternalEmoji},
		{"VoiceConnect", VoiceConnect},
		{"VoiceSpeak", VoiceSpeak},
		{"VoiceVideo", VoiceVideo},
		{"VoiceScreenshare", VoiceScreenshare},
		{"VoicePTT", VoicePTT},
		{"VoiceMuteMembers", VoiceMuteMembers},
		{"VoiceDeafenMembers", VoiceDeafenMembers},
		{"VoiceMoveMembers", VoiceMoveMembers},
		{"VoicePriority", VoicePriority},
		{"ManageAutomod", ManageAutomod},
		{"ViewReports", ViewReports},
		{"ManageReports", ManageReports},
		{"SlowmodeExempt", SlowmodeExempt},
	}

	for i := 0; i < len(perms); i++ {
		for j := i + 1; j < len(perms); j++ {
			if perms[i].perm&perms[j].perm != 0 {
				t.Errorf("%s and %s share bits", perms[i].name, perms[j].name)
			}
		}
	}
}

func TestDenyWinsPattern(t *testing.T) {
	base := ViewChannels | SendMessages | AddReactions

	roleAllow := ManageMessages
	roleDeny := SendMessages

	result := base.Add(roleAllow).Remove(roleDeny)

	if result.Has(SendMessages) {
		t.Error("SendMessages should be denied after deny override")
	}
	if !result.Has(ViewChannels) {
		t.Error("ViewChannels should remain after deny of unrelated permission")
	}
	if !result.Has(ManageMessages) {
		t.Error("ManageMessages should be added by role allow")
	}
	if !result.Has(AddReactions) {
		t.Error("AddReactions should remain untouched")
	}
}
